File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):    
    start, end = 0, max(times)*n
    answer = end
    
    while start <= end:
        mid = (start + end) // 2
        
        people = 0
        for time in times:
            people += mid // time
        
        if people >= n:
            end = mid - 1
            if answer > mid:
                answer = mid
            
        elif people < n:
            start = mid + 1
    
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;
class Solution {
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        long answer = Long.MAX_VALUE;
        long start = 0;
        long end = (long)times[times.length-1]*(long)n;
        while (start <= end) {
            long mid = (start + end) / 2;
            long total = 0;
            for (int i = 0; i < times.length; i++) {
                total += mid / times[i];
            }
            if (total < n) {
                start = mid + 1;
            } else {
                answer = Math.min(answer, mid);
                end = mid - 1;
            }
        }
        return answer;
    }
}